src/amb_kf.c

```diff
@@ -38,75 +43,102 @@
+ * \param state_dim The dimension of the KF state.
+ * \param h         A row of the observation matrix.
+ * \param R         The variance of the observation.
+ * \param U         KF state covariance U (not packed form)
+ * \param D         KF state covariance D (stored as a vector)
+ * \param f         U^T * h.
+ * \param g         diag(D) * f.
+ * \return          alpha = f * g + R, the innovation variance.
  */
-static void incorporate_scalar_measurement(u32 state_dim, double *h, double R,
-                                           double *U, double *D, double *k)
+static double compute_innovation_terms(u32 state_dim, const double *h,
+                                       double R, const double *U,
+                                       const double *D, double *f, double *g)
 {
-  DEBUG_ENTRY();
-
-  if (DEBUG) {
-    VEC_PRINTF(h, state_dim);
-    printf("R = %.16f", R);
-    if (abs(R) == 0) {
-      printf(" \t (R == 0 exactly)\n");
-    }
-    else {
-      printf("\n");
-    }
-  }
-
-  double f[state_dim]; /*  f = U^T * h. */
   memcpy(f, h, state_dim * sizeof(double));
+  /*  f = U^T * h. */
   cblas_dtrmv(CblasRowMajor, CblasUpper, CblasTrans, CblasUnit,
               /* ^ CBLAS_ORDER, CBLAS_UPLO, CBLAS_TRANSPOSE transA, CBLAS_DIAG. */
               state_dim, U,     /* int N, double *A. */
               state_dim, f, 1); /*  int lda, double *X, int incX. */
 
 
-  double g[state_dim]; /*  g = diag(D) * f. */
-  double alpha = R;    /*  alpha = f * g + R = f^T * diag(D) * f + R. */
+  /*  g = diag(D) * f.
+      alpha = f * g + R = f^T * diag(D) * f + R. */
+  double alpha = R;
   for (u32 i=0; i<state_dim; i++) {
     g[i] = D[i] * f[i];
     alpha += f[i] * g[i];
   }
-  if (DEBUG) {
-    VEC_PRINTF(f, state_dim);
-    VEC_PRINTF(g, state_dim);
-    printf("alpha = %.16f", alpha);
-    if (abs(alpha) == 0) {
-      printf(" \t (alpha == 0 exactly)\n");
-    }
-    else {
-      printf("\n");
-    }
+  return alpha;
+}
+
+
+/** In place updating of the state cov and k vec using a scalar observation
+ * This is from section 10.2.1 of Gibbs [1], with some extra logic for handling
+ * singular matrices, dictating that zeros from cov_D dominate in a particular
+ * potential 0 / 0.
+ * We also make it more robust, by multiplying k by k_scalar <=  1.
+ * To still perform the update in factored form, we modified the derivation
+ * in section V.3 of [2] to arrive at the following code.
+ *
+ * \param kf        The KF to update
+ * \param R         The measurement variance
+ * \param f         U^T * h
+ * \param g         diag(D) * f
+ * \param alpha     The innovation variance
+ * \param k_scalar  A scalar to multiply the Kalman gain by (softens outliers)
+ * \param innov     The difference between the actual and predicted observation
+ */
+static void update_kf_state(nkf_t *kf, double R, double *f, double *g,
+                            double alpha, double k_scalar,
+                            double innov)
+{
+  DEBUG_ENTRY();
+
```

Comment:
@fnoble what happens with `assert`s in the firmware. Are they noop's for the actual firmware artifact? In any case, wouldn't we want to have broader checking (like with the first check on `k_scalar`) with return values on errors?